src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesConsumer.java

```diff
@@ -127,142 +135,254 @@ public void addKNNBinaryField(FieldInfo field, DocValuesProducer valuesProducer,
-            indexCreator = () -> createKNNIndexFromScratch(field, pair, knnEngine, indexPath);
-        }
-
-        // Skip index creation if no vectors or docs in segment
-        if (pair.getVectorAddress() == 0 || pair.docs.length == 0) {
-            logger.info("Skipping engine index creation as there are no vectors or docs in the segment");
-            return;
-        }
-
-        long arraySize = calculateArraySize(pair.docs.length, pair.getDimension(), pair.serializationMode);
-
-        if (isMerge) {
-            KNNGraphValue.MERGE_CURRENT_OPERATIONS.increment();
-            KNNGraphValue.MERGE_CURRENT_DOCS.incrementBy(pair.docs.length);
-            KNNGraphValue.MERGE_CURRENT_SIZE_IN_BYTES.incrementBy(arraySize);
-            recordMergeStats(pair.docs.length, arraySize);
-        }
-
-        // Increment counter for number of graph index requests
-        KNNCounter.GRAPH_INDEX_REQUESTS.increment();
+        state.directory.createOutput(engineFileName, state.context).close();
+        boolean fromScratch = !field.attributes().containsKey(MODEL_ID);
+        boolean iterative = fromScratch && KNNEngine.FAISS == knnEngine;
+        createKNNIndex(field, values, knnEngine, indexPath, fromScratch, iterative, isMerge);
 
         if (isRefresh) {
             recordRefreshStats();
         }
-
-        // Ensure engineFileName is added to the tracked files by Lucene's TrackingDirectoryWrapper
-        state.directory.createOutput(engineFileName, state.context).close();
-        indexCreator.createIndex();
         writeFooter(indexPath, engineFileName);
     }
 
+    private void currentMergeStats(int length, long arraySize) {
+        KNNGraphValue.MERGE_CURRENT_OPERATIONS.increment();
+        KNNGraphValue.MERGE_CURRENT_DOCS.incrementBy(length);
+        KNNGraphValue.MERGE_CURRENT_SIZE_IN_BYTES.incrementBy(arraySize);
+        KNNGraphValue.MERGE_TOTAL_OPERATIONS.increment();
+        KNNGraphValue.MERGE_TOTAL_DOCS.incrementBy(length);
+        KNNGraphValue.MERGE_TOTAL_SIZE_IN_BYTES.incrementBy(arraySize);
+    }
+
     private void recordMergeStats(int length, long arraySize) {
         KNNGraphValue.MERGE_CURRENT_OPERATIONS.decrement();
         KNNGraphValue.MERGE_CURRENT_DOCS.decrementBy(length);
         KNNGraphValue.MERGE_CURRENT_SIZE_IN_BYTES.decrementBy(arraySize);
-        KNNGraphValue.MERGE_TOTAL_OPERATIONS.increment();
-        KNNGraphValue.MERGE_TOTAL_DOCS.incrementBy(length);
-        KNNGraphValue.MERGE_TOTAL_SIZE_IN_BYTES.incrementBy(arraySize);
     }
 
     private void recordRefreshStats() {
         KNNGraphValue.REFRESH_TOTAL_OPERATIONS.increment();
     }
 
-    private void createKNNIndexFromTemplate(Model model, KNNCodecUtil.Pair pair, KNNEngine knnEngine, String indexPath) {
+    private Map<String, Object> genParameters(boolean fromScratch, FieldInfo fieldInfo, KNNEngine knnEngine) throws IOException {
         Map<String, Object> parameters = new HashMap<>();
+        ;
+        if (fromScratch) {
```

Comment:
Can we just have branch between fromScratch and not during index initialization? Then, everything else should proceed the same.

Reply:
I don't know if I'm following. You mean have two different functions, one for fromScratch, one for fromTemplate?